app/uk/gov/hmrc/helptosavefrontend/repo/SessionStore.scala

```diff
@@ -0,0 +1,139 @@
+/*
+ * Copyright 2018 HM Revenue & Customs
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package uk.gov.hmrc.helptosavefrontend.repo
+
+import cats.data.{EitherT, OptionT}
+import cats.instances.either._
+import cats.instances.future._
+import cats.syntax.either._
+import com.google.inject.{ImplementedBy, Inject, Singleton}
+import play.api.libs.json.{JsValue, Json, Reads, Writes}
+import play.modules.reactivemongo.ReactiveMongoComponent
+import uk.gov.hmrc.cache.model.Id
+import uk.gov.hmrc.cache.repository.CacheMongoRepository
+import uk.gov.hmrc.helptosavefrontend.config.FrontendAppConfig
+import uk.gov.hmrc.helptosavefrontend.metrics.Metrics
+import uk.gov.hmrc.helptosavefrontend.models.HTSSession
+import uk.gov.hmrc.helptosavefrontend.util.{Result, toFuture}
+import uk.gov.hmrc.http.HeaderCarrier
+
+import scala.concurrent.{ExecutionContext, Future}
+
+@ImplementedBy(classOf[SessionStoreImpl])
+trait SessionStore {
+
+  def get(implicit reads: Reads[HTSSession], hc: HeaderCarrier): Result[Option[HTSSession]]
+
+  def store(body: HTSSession)(implicit writes: Writes[HTSSession], hc: HeaderCarrier): Result[Unit]
+}
+
+@Singleton
+class SessionStoreImpl @Inject() (mongo: ReactiveMongoComponent, metrics: Metrics)(implicit appConfig: FrontendAppConfig, ec: ExecutionContext) extends SessionStore {
+
+  private val expireAfterSeconds = appConfig.mongoSessionExpireAfter.toSeconds
+
+  private lazy val cacheRepository = new CacheMongoRepository("sessions", expireAfterSeconds)(mongo.mongoConnector.db, ec)
+
+  private type EitherStringOr[A] = Either[String, A]
+
+  override def get(implicit reads: Reads[HTSSession], hc: HeaderCarrier): Result[Option[HTSSession]] = {
+
+    val timerContext = metrics.sessionReadTimer.time()
```

Comment:
this timer should start just before `doFindById` below - if there is no session id there is no need to run the timer